presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,654 @@
+            line.setStart(startPoint);
+            line.setEnd(endPoint);
+            double weight = line.calculateLength2D();
+            weightSum += weight;
+            xSum += (startPoint.getX() + endPoint.getX()) * weight / 2;
+            ySum += (startPoint.getY() + endPoint.getY()) * weight / 2;
+        }
+        return new Point(xSum / weightSum, ySum / weightSum);
+    }
+
+    // Polygon centroid: area weighted average of centroids in case of holes
+    private static Point getPolygonCentroid(OGCGeometry geometry)
+    {
+        Polygon polygon = (Polygon) geometry.getEsriGeometry();
+        final int pathCount = polygon.getPathCount();
+
+        if (pathCount == 1) {
+            return getPolygonSansHolesCentroid(polygon);
+        }
+
+        double xSum = 0;
+        double ySum = 0;
+        double areaSum = 0;
+
+        for (int i = 0; i < pathCount; i++) {
+            final int startIndex = polygon.getPathStart(i);
+            final int endIndex = polygon.getPathEnd(i);
+
+            final Polygon sansHoles = getPolygon(polygon, startIndex, endIndex);
+
+            final Point centroid = getPolygonSansHolesCentroid(sansHoles);
+            final double area = sansHoles.calculateArea2D();
+
+            xSum += centroid.getX() * area;
+            ySum += centroid.getY() * area;
+            areaSum += area;
+        }
+
+        return new Point(xSum / areaSum, ySum / areaSum);
+    }
+
+    private static Polygon getPolygon(Polygon polygon, int startIndex, int endIndex)
+    {
+        final Polyline boundary = new Polyline();
+        boundary.startPath(polygon.getPoint(startIndex));
+        for (int i = startIndex + 1; i < endIndex; i++) {
+            Point current = polygon.getPoint(i);
+            boundary.lineTo(current);
+        }
+
+        final Polygon newPolygon = new Polygon();
+        newPolygon.add(boundary, false);
+        return newPolygon;
+    }
+
+    // Polygon sans holes centroid:
+    // c[x] = (Sigma(x[i] + x[i + 1]) * (x[i] * y[i + 1] - x[i + 1] * y[i]), for i = 0 to N - 1) / (6 * signedArea)
+    // c[y] = (Sigma(y[i] + y[i + 1]) * (x[i] * y[i + 1] - x[i + 1] * y[i]), for i = 0 to N - 1) / (6 * signedArea)
+    private static Point getPolygonSansHolesCentroid(Polygon polygon)
+    {
+        final int pointCount = polygon.getPointCount();
+        double xSum = 0;
+        double ySum = 0;
+        double signedArea = 0;
+        for (int i = 0; i < pointCount; i++) {
+            Point current = polygon.getPoint(i);
+            Point next = (i + 1 == polygon.getPointCount()) ? polygon.getPoint(0) : polygon.getPoint(i + 1);
+            double ladder = current.getX() * next.getY() - next.getX() * current.getY();
+            xSum += (current.getX() + next.getX()) * ladder;
+            ySum += (current.getY() + next.getY()) * ladder;
+            signedArea += ladder;
+        }
+        return new Point(xSum / (signedArea * 3), ySum / (signedArea * 3));
```

Comment:
The comments before the method say "**6** * signedArea", but this multiplies by 3. Which one is it?

Reply:
it is 6. sum of ladder actually is 2 * signedArea, I merged computation there. Will update